inginious/frontend/task_dispensers/toc.py

```diff
@@ -26,6 +27,34 @@ def get_name(cls, language):
         """ Returns the localized task dispenser name """
         return _("Table of contents")
 
+    def get_course_grade(self, username, course, user_task_list):
+        tasks = course.get_tasks()
```

Comment:
Here you should remove the dependency to course and use the internal ``task_list_func()`` call. You can take the ``get_user_task_list`` implementation as an example, and even, as this would be way cleaner, use that internal method to avoid passing the list, that, in fine, is generated by the task dispenser, as argument.

This will better respect the encapsulation principle. 

This remark is also valid for ``combinatory_test`` code.